File: apps/backend/src/contracts/session.ts
/** Session — the central object for one teaching session and its state (§6.2). */

import { z } from "zod";
import { sessionStatusSchema } from "./enums.js";

export const sessionSchema = z.object({
  sessionId: z.string(),
  topicId: z.string(),
  status: sessionStatusSchema,
  createdAt: z.string(),
  startedAt: z.string().optional(),
  endedAt: z.string().optional(),
  turnCount: z.number().int().default(0),
  /** Total LLM tokens (input + output) this session has spent so far. */
  tokensUsed: z.number().int().default(0),
  /** The most recent evaluation (the latest round). */
  evaluationId: z.string().optional(),
  /**
   * All evaluations for this session, oldest first — one per ended round
   * (resume extension to §6.2). The last entry equals `evaluationId`.
   */
  evaluationIds: z.array(z.string()).default([]),
});

export type Session = z.infer<typeof sessionSchema>;

File: apps/backend/app/contracts/session.py
"""Session — the central object for one teaching session and its state (§6.2)."""

from __future__ import annotations

from pydantic import Field

from .base import CamelModel
from .enums import SessionStatus


class Session(CamelModel):
    session_id: str = Field(description="Unique session identifier")
    topic_id: str = Field(description="Reference to the Topic being taught")
    status: SessionStatus = Field(
        description="SETUP | TEACHING | ENDED | EVALUATED"
    )
    created_at: str = Field(description="Session creation time (ISO-8601)")
    started_at: str | None = Field(default=None, description="Time teaching started")
    ended_at: str | None = Field(default=None, description="Time the session ended")
    turn_count: int = Field(
        default=0, description="Number of teaching turns taken so far"
    )
    evaluation_id: str | None = Field(
        default=None, description="Reference to the EvaluationResult once it exists"
    )
